zepa.go

```
@@ -15,38 +17,170 @@ const (
 
+func (m *Machine) mv(inst Instruction) {}
+
+func (m *Machine) add(inst Instruction) {}
+
+func (m *Machine) sub(inst Instruction) {}
+
+func (m *Machine) cmp(inst Instruction) {}
+
+func (m *Machine) jump(inst Instruction) {}
+
+func (m *Machine) load(inst Instruction) {}
+
+func (m *Machine) store(inst Instruction) {}
+
 func (m *Machine) fetch() {
-	currentInstructionAddress := m.registers[pc]
-	currentInstruction := m.memory[currentInstructionAddress]
-	// Update Instruction Register with current instruction
-	m.registers[ir] = currentInstruction
-	// Increment Program Counter
-	m.registers[pc] += 1
+	var completeInstruction uint32 = 0
+	for i := 0; i < 4; i++ {
+		currentInstructionAddress := m.registers[pc]
+		currentInstruction := m.memory[currentInstructionAddress]
+		completeInstruction = completeInstruction | uint32(currentInstruction)<<(24-8*i)
+		m.registers[pc] += 1
+	}
+	m.registers[ir] = completeInstruction
 }
 
-func (m *Machine) decode() {}
+func (m *Machine) decodeRTypeInst(instruction uint32) Instruction {
+	offsetOpcode := word - opcodeLength
+	offSetRd := offsetOpcode - rdLength
+	offSetRs1 := offSetRd - rs1Length
+	offSetRs2 := offSetRs1 - rs2Length
+	offSetFunct5 := offSetRs2 - funct5Length
+	offSetFunct6 := offSetFunct5 - funct6Length
+
+	opcode := instruction >> (uint32(offsetOpcode)) & 0b111111
+	rd := (instruction >> uint32(offSetRd)) & registerBitMask
+	rs1 := (instruction >> (uint32(offSetRs1))) & registerBitMask
+	rs2 := (instruction >> (uint32(offSetRs2))) & registerBitMask
+	funct5 := (instruction >> (uint32(offSetFunct5))) & funct5BitMask
+	funct6 := (instruction >> (uint32(offSetFunct6))) & funct6BitMask
+
+	operation := operations[byte(opcode)]
+
+	return Instruction{
+		opcode: operation,
+		rd:     Register(rd),
+		rs1:    Register(rs1),
+		rs2:    Register(rs2),
+		funct5: byte(funct5),
+		funct6: byte(funct6),
+	}
+}
+
+func (m *Machine) decodeITypeInst(instruction uint32) Instruction {
+	offsetOpcode := word - opcodeLength
```

Should return an error checking the operands